climada/util/api_client.py

```diff
@@ -0,0 +1,472 @@
+        params = {'data_type_group': data_type_group} \
+            if data_type_group else {}
+        return [DataTypeInfo(**jobj) for jobj in Client._request_200(url, params=params)]
+
+    def get_data_type(self, data_type):
+        """Returns the data type from the climada data API
+        with a given name.
+
+        Parameters
+        ----------
+        data_type : str
+            data type name
+
+        Returns
+        -------
+        DataTypeInfo
+
+        Raises
+        ------
+        NoResult
+            if there is no such data type registered
+        """
+        url = f'{self.host}/rest/data_type/{quote(data_type)}'
+        return DataTypeInfo(**Client._request_200(url))
+
+    def _download(self, url, path, replace=False):
+        """Downloads a file from the given url to a specified location.
+
+        Parameters
+        ----------
+        url : str
+            the link to the file to be downloaded
+        path : Path
+            download path, if it's a directory the original file name is kept
+        replace : bool, optional
+            flag to indicate whether a present file with the same name should
+            be replaced
+
+        Returns
+        -------
+        Path
+            Path to the downloaded file
+
+        Raises
+        ------
+        FileExistsError
+            in case there is already a file present at the given location
+            and replace is False
+        """
+        if path.is_dir():
+            path /= unquote(url.split('/')[-1])
+        if path.is_file() and not replace:
+            raise FileExistsError(path)
+        with requests.get(url, stream=True) as stream:
+            stream.raise_for_status()
+            with open(path, 'wb') as dump:
+                for chunk in stream.iter_content(chunk_size=self.chunk_size):
+                    dump.write(chunk)
+        return path
+
+    def _tracked_download(self, remote_url, local_path):
+        if local_path.is_dir():
+            raise Exception("tracked download requires a path to a file not a directory")
+        path_as_str = str(local_path.absolute())
+        try:
+            dlf = Download.create(url=remote_url, path=path_as_str, startdownload=datetime.utcnow())
+        except IntegrityError as ierr:
+            dlf = Download.get(Download.path==path_as_str)
+            if dlf.url != remote_url:
+                raise Exception("this file has been downloaded from another url, "
+                    "please purge the entry from data base before trying again") from ierr
+            return dlf
+        try:
+            self._download(url=remote_url, path=local_path, replace=True)
+            dlf.enddownload = datetime.utcnow()
+            dlf.save()
+        except Exception:
+            dlf.delete_instance()
+            raise
+        return Download.get(Download.path==path_as_str)
+
+    def download_file(self, local_path, fileinfo, check=checksize, retries=3):
```

Comment:
this one could also have a automatically configured path as in download_dataset based on the datatype,... that it belongs to?

Reply:
Yes, but not so easy: all information on how to automatically build the path is in dataset_info, which is not available.
To include such an option would load a lot of unessential complexity upon `download_file` which I want to avoid.